python/pfs/drp/stella/lsf.py

```diff
@@ -0,0 +1,688 @@
+        """Get value(s) from the kernel
+
+        Supports integer or array indexing.
+
+        Values off the end are zero.
+        """
+        if isinstance(index, int):
+            return self._getitem(index)
+        return np.vectorize(self._getitem)(index)
+
+    def __setitem__(self, index, value):
+        """Set value(s) in the kernel.
+
+        Supports integer or array indexing.
+
+        Values off the end raise an `IndexError`.
+        """
+        return np.vectorize(self._setitem)(index, value)
+
+    def __iter__(self):
+        """Iterator
+
+        Produces ``index,value`` pairs.
+        """
+        return zip(iter(self.indices), iter(self.values))
+
+    def _getOtherValues(self, other):
+        """Get values from another Kernel1D IFF dimensions match"""
+        if isinstance(other, Kernel1D):
+            if self.min != other.min or self.max != other.max:
+                raise IndexError("Kernel1D ranges don't match: %d..%d vs %d..%d" %
+                                 (self.min, self.max, other.min, other.max))
+            other = other.values
+        return other
+
+    def __eq__(self, other):
+        """Test for equality"""
+        return (self.__class__ == other.__class__ and
+                self.min == other.min and
+                self.max == other.max and
+                np.all(self.values == other.values))
+
+    def __imul__(self, other):
+        """Multiplication in-place
+
+        Supports multiplication by a scalar or an array.
+        """
+        self.values *= self._getOtherValues(other)
+        return self
+
+    def __itruediv__(self, other):
+        """Division in-place
+
+        Supports division by a scalar or an array.
+        """
+        self.values /= self._getOtherValues(other)
+        return self
+
+    def __reduce__(self):
+        """Pickle prescription"""
+        return self.__class__, (self.values, self.center, False)
+
+    def __repr__(self):
+        """Representation"""
+        return "%s(%s, %f)" % (self.__class__.__name__, self.values, self.center)
+
+    def normalization(self):
+        """Return the normalization"""
+        return np.sum(self.values)
+
+    def convolve(self, array):
+        """Convolve an array by the kernel
+
+        The convolved array has the same length as the input array in the usual
+        case that the array is longer than the kernel.
+
+        Note that the ends of the array will not be fully convolved.
+
+        Parameters
+        ----------
+        array : `numpy.ndarray`, floating-point
+            Array to convolve.
+
+        Raises
+        ------
+        RuntimeError
+            If the array is shorter than the kernel.
+        """
+        if len(array) < self.length:
+            raise RuntimeError("Array too short for convolution by this kernel")
+        if self.center != (self.length - 1)//2:
+            # Pad to put center of kernel in the middle, so we don't introduce a shift
```

Comment:
A gross shift.  If the kernel has a non-zero first moment around the centre it will still shift and that will be Very Bad.  Probably should check in the ctor